staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -659,59 +680,80 @@ func TestContextCancel(t *testing.T) {
 		t.Fatal(err)
 	}
 	qs := qsc.Complete(fq.DispatchingConfig{ConcurrencyLimit: 1})
-	counter.Add(1) // account for the goroutine running this test
+	counter.Add(1) // account for main activity of the goroutine running this test
 	ctx1 := context.Background()
-	b2i := map[bool]int{false: 0, true: 1}
-	var qnc [2][2]int32
-	req1, _ := qs.StartRequest(ctx1, &fcrequest.WorkEstimate{Seats: 1}, 1, "", "fs1", "test", "one", func(inQueue bool) { atomic.AddInt32(&qnc[0][b2i[inQueue]], 1) })
+	pZero := func() *int32 { var zero int32; return &zero }
+	// counts of calls to the QueueNoteFns
+	queueNoteCounts := map[int]map[bool]*int32{
+		1: {false: pZero(), true: pZero()},
+		2: {false: pZero(), true: pZero()},
+	}
+	queueNoteFn := func(fn int) func(inQueue bool) {
+		return func(inQueue bool) { atomic.AddInt32(queueNoteCounts[fn][inQueue], 1) }
+	}
+	fatalErrs := []string{}
+	var errsLock sync.Mutex
+	expectQNCount := func(fn int, inQueue bool, expect int32) {
+		if a := atomic.LoadInt32(queueNoteCounts[fn][inQueue]); a != expect {
+			errsLock.Lock()
+			defer errsLock.Unlock()
+			fatalErrs = append(fatalErrs, fmt.Sprintf("Got %d calls to queueNoteFn%d(%v), expected %d", a, fn, inQueue, expect))
+		}
+	}
+	expectQNCounts := func(fn int, expectF, expectT int32) {
+		expectQNCount(fn, false, expectF)
+		expectQNCount(fn, true, expectT)
+	}
+	req1, _ := qs.StartRequest(ctx1, &fcrequest.WorkEstimate{Seats: 1}, 1, "", "fs1", "test", "one", queueNoteFn(1))
 	if req1 == nil {
 		t.Error("Request rejected")
 		return
 	}
-	if a := atomic.AddInt32(&qnc[0][0], 0); a != 1 {
-		t.Errorf("Got %d calls to queueNoteFn1(false), expected 1", a)
-	}
-	if a := atomic.AddInt32(&qnc[0][1], 0); a != 1 {
-		t.Errorf("Got %d calls to queueNoteFn1(true), expected 1", a)
-	}
-	var executed1 bool
-	idle1 := req1.Finish(func() {
-		executed1 = true
-		ctx2, cancel2 := context.WithCancel(context.Background())
-		tBefore := time.Now()
-		go func() {
-			time.Sleep(time.Second)
-			if a := atomic.AddInt32(&qnc[1][0], 0); a != 0 {
-				t.Errorf("Got %d calls to queueNoteFn2(false), expected 0", a)
-			}
-			if a := atomic.AddInt32(&qnc[1][1], 0); a != 1 {
-				t.Errorf("Got %d calls to queueNoteFn2(true), expected 1", a)
+	expectQNCounts(1, 1, 1)
+	var executed1, idle1 bool
```

Comment:
I think I need to define the variable up front, it is set in a forked goroutine and checked in the main goroutine.  The request processing needs to be in a forked goroutine so that it can run concurrently with the doneCh reaction under management of the fake clock.

Reply:
Sorry - I missed that it's being exercised at the end of the test.